### ISSUE TITLE
PrePARE table memory issue

### ISSUE THREAD

**Issue by glevava** [PCMDI/cmor#314]
Hi @dnadeau4 
I make a fresh install of CMOR 3.3.1 and it seems that PrePARE has a table memory issue.
Here what I done:
- Apply PrePARE to one file => Compliant
- Apply PrePARE sequentially to a directory including the same above file => this file returns CV error.

I observed that the CV error appears for all files checked after the first "table change". For instance:
- File 1 <- tableA = Compliant
- File 2 <- tableA = Compliant
- File 3 <- tableB = Error
- File 4 <- tableC = Error
- etc. with systematic CV error

Each file checked separately returns CMIP6 compliance.

**Comment by glevava** [comment]
The error still appears in a fresh conda install of CMOR 3.3.1 in our HPD env. After further investigation, I confirm a "table memory" side effect in the 'C' internal memory. It seems that the following lines into PrePARE have different behavior depending on the env:
```
# -------------------------------------------------------------------
# call setup() to clean all 'C' internal memory.
# -------------------------------------------------------------------
cmip6_cv.setup(inpath=self.cmip6_table_path, exit_control=cmip6_cv.CMOR_NORMAL)
```
An explicit `cmip6_cv.purge()` or `cmip6_cv_reset()` vacuuming 'C' memory could be helpful here to ensure memory cleaning whatever the env and C libraries version/compilation.

**Comment by glevava** [comment]
Find the solution. This is due to the cmip6_cv.set_CV_Error() which is not reinitialize when checking several files sequentially. I will suggest a PR to solve this.

**Comment by durack1** [comment]
@glevava thanks for the detective work on tracking this down, a PR would very much be appreciated and if all the tests pass should hopefully be merged quickly

**Comment by glevava** [comment]
Hi @durack1 @dnadeau4 I made a [PR](https://github.com/PCMDI/cmor/pull/315) to solve this issue.
The PR also provide some improvements including "OR" operator support from CMOR table for the `cell_measures` attribute.
Thanks

**Comment by taylor13** [comment]
All instances of an "OR" construct are found either in the SImon or SIday table.  The data request should not include cell_measures with an "OR" operator in the cell_measures attribute.  I think this should be corrected by Martin following https://github.com/cmip6dr/CMIP6_DataRequest_VariableDefinitions/issues/252.  It is acceptable that PrePARE can handle this case (until the data request is corrected of possibly it might be needed in the future), but those using CMOR cannot at present correctly handle the sea ice variables where "OR" appears in cell_measures.  So the data request needs correcting.

**Comment by dnadeau4** [comment]
@glevava if the change I made in the PrePARE branch works for you, can you close the issue?

Thanks!

**Comment by glevava** [comment]
of course I will test the latest script version from "PrePARE" branch tomorrow, and close the issue after testing it for sure. Thanks again for your quick merge !

**Comment by glevava** [comment]
Hi @dnadeau4 I tested PrePARE.py version from latest commit on the "PrePARE" branch. And the "table memory" issue raises again due to `cmip6_cv.get_CV_error()` you rollbacked. See my comment here: [https://github.com/PCMDI/cmor/issues/316#issuecomment-373321410](https://github.com/PCMDI/cmor/issues/316#issuecomment-373321410)

I can think about another way to deal with different return codes but please don't use any `cmip6_cv.{set|get}_CV_Error()` method as it cannot purge the error code for a bulk of files scanned.

**Comment by dnadeau4** [comment]
Here is the set_CV_Error function,  it is a direct python call. 

https://github.com/glevava/cmor/blob/5a9153f8f2785ec6a75b0a406a2dbaaf42276f4a/Src/cmor_CV.c#L430-L436

**Comment by glevava** [comment]
Hi @dnadeau4 , let me try to explain clearer what happens here.
Image you submit two files to be checked by PrePARE (file1.nc then file2.nc).
Let's say file1.nc contains a CV error and file2.nc is CMIP6 compliant.

So, we submit both files to PrePARE in a sequential processing (e.g., in one command-line call).
- PrePARE checks file1.nc and find the CV error through one of the `cmip6_cv.check_***()` method. 
- Consequently, `CV_ERROR` is set to `1` in the "C memory"
- If [this PrePARE line](https://github.com/glevava/cmor/blob/5a9153f8f2785ec6a75b0a406a2dbaaf42276f4a/LibCV/PrePARE/PrePARE.py#L610) test the `cmip6_cv.get_CV_error()` it will return `1` and raises the `KeyboardInterrupt ` exception with the error message that the file isn't CMIP6 compliant, as expected.

- Now PrePARE, in the same process,  checks file2.nc and find no CV error.
- But, `CV_ERROR` is still set to `1` in the "C memory" due to the previous file and because of no reset to `CV_ERROR=0` at each iteration.
- Consequently, after checking file1.nc `cmip6_cv.get_CV_error()` will always return `1` and raises non-CMIP6-compliant message.

As conclusion, we can solve this in two ways.
- Handle such CV_ERROR variable directly into PrePARE code instead of `cmor_CV.c`. That's I have done in my latest PR including `self.cv_error` in the `checkCMIP6` class.

**Comment by dnadeau4** [comment]
Thanks!  This is very clear.  I did not understand what you meant by "table memory" before,  I thought the issue was in the `cmip6_cv.load_table()` function.  This misunderstanding happens sometimes with email communication.   

After testing your solution, I will accept your PR.    

I think I will have to make sure that [these functions](https://github.com/glevava/cmor/blob/5a9153f8f2785ec6a75b0a406a2dbaaf42276f4a/LibCV/PrePARE/PrePARE.py#L416-L424)  `return code` or `errors` are checked.  I will also work on the filename check `error` code. 

Thanks for all your help,  it is greatly appreciated.

**Comment by glevava** [comment]
The issue title is not appropriate. In the beginning without investigating further I also though the issue comes from the table loading in a way or another... Sorry for the mistake.

Thanks for merging and checking return codes from cmip6_cv functions  and look forward a fully operational PrePARE 3.3.2 ; ) !

**Comment by dnadeau4** [comment]
I'll try to push it through Wednesday.

**Comment by dnadeau4** [comment]
Fix in 3.3.2
